File: backend/app/schemas/evaluation.py
"""HTTP contracts for evaluation datasets, runs, and comparisons."""

import enum
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class EvalCaseSource(enum.StrEnum):
    SYNTHETIC = "synthetic"
    MANUAL = "manual"


class EvalRunStatus(enum.StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class DatasetCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    version: str = Field(default="v1", min_length=1, max_length=50)
    construction_method: str = Field(default="synthetic_plus_manual", max_length=100)


class ManualCaseInput(BaseModel):
    query: str = Field(min_length=1, max_length=2_000)
    expected_chunk_ids: list[UUID] = Field(min_length=1)
    review_note: str | None = Field(default=None, max_length=2_000)


class ManualCasesRequest(BaseModel):
    cases: list[ManualCaseInput] = Field(min_length=1, max_length=100)


class SyntheticGenerationRequest(BaseModel):
    count: int = Field(default=50, ge=1, le=80)
    seed: int = Field(default=0, ge=0)
    mode: str = Field(default="http", pattern="^(http|offline)$")


class ReviewCaseRequest(BaseModel):
    is_active: bool
    review_note: str | None = Field(default=None, max_length=2_000)


class DatasetStatsResponse(BaseModel):
    total: int
    active: int
    synthetic_total: int
    synthetic_active: int
    manual_active: int
    synthetic_retention_rate: float


class EvalDatasetResponse(BaseModel):
    id: UUID
    name: str
    version: str
    construction_method: str
    case_count: int
    created_at: datetime
    updated_at: datetime
    stats: DatasetStatsResponse


class EvalCaseResponse(BaseModel):
    id: UUID
    dataset_id: UUID
    query: str
    expected_chunk_ids: list[UUID]
    source: EvalCaseSource
    is_active: bool
    review_note: str | None
    created_at: datetime


class SyntheticGenerationAccepted(BaseModel):
    dataset_id: UUID
    status: str
    count: int


class EvalRunCreateRequest(BaseModel):
    dataset_id: UUID
    config_overrides: dict[str, Any] = Field(default_factory=dict)


class EvalRunAccepted(BaseModel):
    run_id: UUID
    status: EvalRunStatus


class EvalRunResponse(BaseModel):
    id: UUID
    dataset_id: UUID
    status: EvalRunStatus
    config_snapshot: dict[str, Any]
    started_at: datetime | None
    finished_at: datetime | None
    summary_metrics: dict[str, Any] | None
    error_message: str | None
    created_at: datetime


class EvalRunListResponse(BaseModel):
    runs: list[EvalRunResponse]


class EvalCompareResponse(BaseModel):
    baseline_run_id: UUID
    current_run_id: UUID
    baseline_summary: dict[str, Any]
    current_summary: dict[str, Any]
    config_diff: dict[str, dict[str, Any]]
    case_diffs: list[dict[str, Any]]
    improved_cases: list[dict[str, Any]]
    regressed_cases: list[dict[str, Any]]

File: frontend/src/api/evaluation.js
import { USE_MOCK, delay, request } from './client'

const mockRun = {
  id: '00000000-0000-0000-0000-000000000101',
  dataset_id: '00000000-0000-0000-0000-000000000201',
  status: 'completed',
  config_snapshot: {
    retrieval_strategy: 'vector',
    top_k: 5,
    fusion: 'none',
    rerank: false,
    embedding_model: 'BAAI/bge-base-zh-v1.5',
    evaluator_version: 'stage2.v1'
  },
  started_at: '2026-08-05T16:00:00Z',
  finished_at: '2026-08-05T16:00:02Z',
  summary_metrics: {
    case_count: 60,
    recall_at_k: 0.72,
    mrr: 0.58,
    ndcg_at_k: 0.63,
    latency_p50_ms: 31.4,
    latency_p95_ms: 68.2,
    attribution_percentages: {
      not_in_kb: 5,
      not_recalled: 18,
      low_rank: 12,
      hit: 60,
      ambiguous: 5
    }
  },
  error_message: null,
  created_at: '2026-08-05T16:00:00Z'
}

const mockDatasets = [
  {
    id: '00000000-0000-0000-0000-000000000201',
    name: 'Stage 2 demo set',
    version: 'v1',
    construction_method: 'synthetic_plus_manual',
    case_count: 60,
    stats: { total: 80, active: 60, synthetic_total: 60, synthetic_active: 40, manual_active: 20, synthetic_retention_rate: 66.7 }
  }
]

export async function getDatasets() {
  if (USE_MOCK) {
    await delay(220)
    return mockDatasets
  }
  return request('/api/eval/datasets')
}

export async function getRuns() {
  if (USE_MOCK) {
    await delay(320)
    return { runs: [mockRun] }
  }
  return request('/api/eval/runs')
}

export async function triggerRun(datasetId, configOverrides = {}) {
  if (USE_MOCK) {
    await delay(260)
    return { run_id: mockRun.id, status: 'queued' }
  }
  return request('/api/eval/runs', {
    method: 'POST',
    body: { dataset_id: datasetId, config_overrides: configOverrides }
  })
}

export async function compareRuns(currentId, baselineId) {
  if (USE_MOCK) {
    await delay(280)
    return {
      baseline_run_id: baselineId,
      current_run_id: currentId,
      baseline_summary: { recall_at_k: 0.68, mrr: 0.54, ndcg_at_k: 0.59 },
      current_summary: mockRun.summary_metrics,
      config_diff: { evaluator_version: { baseline: 'stage2.v1', current: 'stage2.v1' } },
      case_diffs: [],
      improved_cases: [],
      regressed_cases: []
    }
  }
  return request(`/api/eval/runs/${currentId}/compare?baseline=${baselineId}`)
}
